### ISSUE TITLE
Typechecker: Union of 2 shapes should result in the union of each key?

### ISSUE THREAD

**Issue by eryi** [facebook/hhvm#7790]
### HHVM Version
3.19.1

### Standalone code, or other way to reproduce the problem
````
<?hh //strict
function foo(bool $x):void {
  if ($x) {
    $y = shape('key1' => 2);
  } else {
    $y = shape('key2' => 1);
  }

  if (Shapes::keyExists($y, 'key1')) var_dump($y['key1']);
}
````

### Expected result
hh_client: No errors!

### Actual result
hh_client: The field key1 is undefined

**Comment by mofarrell** [comment]
I think this may be the intended behavior.  https://docs.hhvm.com/hack/shapes/subtyping
`Shapes::keyExists` fails if the key is not known to be defined.  You can most likely get around this by turning the shape into an array, checking for the key, and then using [invariant](https://docs.hhvm.com/hack/types/refining#invariant) to assert its type.

**Comment by eryi** [comment]
The issue with turning shapes into arrays is that you lose type information on every key and it will require a lot of invariants to assert back everything. Better solutions might be:

1. Union each key of each shape (not the best idea now that I think about it)
2. Allow refining using Shapes::keyExists or something like [what Flow does](https://flow.org/blog/2015/07/03/Disjoint-Unions/)

**Comment by eryi** [comment]
Fixed in 3.22.0
